File: leetcodeGo/sliding_window/[blind 75]_LongestSubstringWithoutRepeatingCharacters.go
func lengthOfLongestSubstring(s string) int {
    l,r,longest := 0,0,0
    tracker := make(map[byte]byte)
    for {
        if r >= len(s) {
            break
        }
        for {
            if _, present := tracker[s[r]]; !present {
                break
            } else {
                delete(tracker, s[l])
                l += 1
            }
        }
        tracker[s[r]] = s[r]
        if longest < r - l + 1 {
            longest = r - l + 1
        }
        r += 1
    }
    return longest
}

//TC: O(n)
//SC: O(n) - for the hashset

File: leetcodeJS/sliding_window/[blind 75]_LongestSubstringWithoutRepeatingCharacters.ts
lengthOfLongestSubstring(s) {
    let longest = 0;
    let tracker = new Set();
    let [l, r] = [0,0];
    while (r < s.length) {
        while (tracker.has(s[r])) {
            tracker.delete(s[l]);
            l += 1;
        }
        tracker.add(s[r]);
        longest = Math.max(longest, r-l+1);
        r += 1;
    }
    return longest;
}

//TC: O(n)
//SC: O(n) - for tracker 
